### PR TITLE
Task/android12/sdk 1067(SDK-1067)

### PR DESCRIPTION
**Pull request by william-ct**
* Adds support for [apps targeting Android 12 (API 31)](https://developer.android.com/about/versions/12/behavior-changes-12)
  This version is compatible with all new Android 12 changes like Notification Trampolines, Pending Intents Mutability and Safer Component Exporting.
  For more information check out the [CleverTap documentation for Android 12 here](https://developer.clevertap.com/docs/android-12-changes)
* Deprecated `CTPushNotificationReceiver` and `CTNotificationIntentService` as a part of Notification Trampoline restrictions in Android 12 (API 31)
* Last version with support for custom FCM Sender ID for generating the FCM token
* Fixes a bug where `UTM Visited` event was not being raised on click of a direct deep link

### CODE REVIEW

File: clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/CTInAppBaseFullNativeFragment.java [hunk 1]
```diff
@@ -1,27 +1,39 @@
...
 import android.util.DisplayMetrics;
 import android.view.View;
 import android.view.WindowManager;
 import android.widget.Button;
 
+import com.clevertap.android.sdk.Logger;
+
 public abstract class CTInAppBaseFullNativeFragment extends CTInAppBaseFullFragment {
 
     int getDPI() {
```
[the remark this answers is not in the document]
Comment (darshanclevertap) [reply clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/CTInAppBaseFullNativeFragment.java:21]: This is done

File: clevertap-core/src/main/java/com/clevertap/android/sdk/pushnotification/CTNotificationIntentService.java [hunk 2]
```diff
@@ -9,6 +10,10 @@
 import com.clevertap.android.sdk.Logger;
 import com.clevertap.android.sdk.Utils;
 
+/**
+ * In order to support Android 12 notification trampoline, this class is being deprecated from 4.3.0 onwards
```
[the remark this answers is not in the document]
Comment (darshanclevertap) [reply clevertap-core/src/main/java/com/clevertap/android/sdk/pushnotification/CTNotificationIntentService.java]: this is done

File: clevertap-core/src/main/java/com/clevertap/android/sdk/pushnotification/PushProviders.java [hunk 3]
```diff
@@ -893,18 +931,53 @@ private void setPingFrequency(Context context, int pingFrequency) {
         StorageHelper.putInt(context, Constants.PING_FREQUENCY, pingFrequency);
     }
 
+    private void createAlarmScheduler(Context context) {
```
[the remark this answers is not in the document]
Comment (darshanclevertap) [reply clevertap-core/src/main/java/com/clevertap/android/sdk/pushnotification/PushProviders.java:887]: this is done

File: clevertap-core/src/main/java/com/clevertap/android/sdk/pushnotification/PushProviders.java [hunk 4]
```diff
@@ -1086,7 +1150,7 @@ private void triggerNotification(Context context, Bundle extras, String notifMes
 
         nb.setContentTitle(notifTitle)
                 .setContentText(notifMessage)
-                .setContentIntent(pIntent)
+                .setContentIntent(LaunchPendingIntentFactory.getLaunchPendingIntent(extras, context))
```
[the remark this answers is not in the document]
Comment (darshanclevertap) [reply clevertap-core/src/main/java/com/clevertap/android/sdk/pushnotification/PushProviders.java:1106]: This is done

File: sample/build.gradle [hunk 5]
```diff
@@ -63,27 +63,27 @@ dependencies {
     //implementation 'com.clevertap.android:clevertap-hms-sdk:1.0.0'
     //implementation 'com.huawei.hms:push:5.0.0.300'
```
[the remark this answers is not in the document]
Comment (darshanclevertap) [reply sample/build.gradle]: This is done

File: sample/build.gradle [hunk 6]
```diff
@@ -63,27 +63,27 @@ dependencies {
...
 
-    implementation 'com.google.android.gms:play-services-location:17.0.0'// Needed for geofence
-    implementation 'androidx.work:work-runtime:2.3.4'// Needed for geofence
-    implementation 'androidx.concurrent:concurrent-futures:1.0.0'// Needed for geofence
+    implementation 'com.google.android.gms:play-services-location:18.0.0'// Needed for geofence
+    implementation 'androidx.work:work-runtime:2.7.0-rc01'// Needed for geofence
+    implementation 'androidx.concurrent:concurrent-futures:1.1.0'// Needed for geofence
 
     implementation 'com.google.firebase:firebase-messaging:20.2.4' //Needed for FCM
     implementation 'com.google.android.gms:play-services-ads:19.4.0' //Needed to use Google Ad Ids
```
[the remark this answers is not in the document]
Comment (darshanclevertap) [reply sample/build.gradle]: This is done

File: versions.properties [hunk 7]
```diff
@@ -88,37 +173,276 @@ version.org.powermock..powermock-module-junit4=2.0.9
...
+##        # available=2.7.0-alpha03
+##        # available=2.7.0-alpha04
+##        # available=2.7.0-alpha05
+##        # available=2.7.0-beta01
+##        # available=2.7.0-rc01
+version.com.android.installreferrer..installreferrer=2.2
+##                                       # available=2.2
 version.com.android.tools.lint..lint-api=27.0.1
 version.com.android.tools.lint..lint-checks=27.0.1
 version.com.clevertap.android..clevertap-android-sdk=4.2.0
```
[the remark this answers is not in the document]
Comment (darshanclevertap) [reply versions.properties]: This is done

**Review by darshanclevertap — APPROVED** [review]
LGTM 👍🏾

**Review by piyush-kukadiya — APPROVED** [review]
LGTM

**Pull request merged**